karapace/backup/v3/schema.proto

```diff
@@ -0,0 +1,140 @@
+  // [SemVer](https://semver.org/) compatible.
+  string tool_version = 2;
+
+  // The UNIX time the backup creation started.
+  int64 create_time = 3;
+
+  // The estimated record count included in this backup. This is just an
+  // estimate, because it is not possible to determine the actual record count
+  // upfront when consuming from a topic based just on the start and end
+  // offsets. The estimation is still useful for getting an idea how many
+  // records there are, e.g. to create a progress bar.
+  uint64 estimated_record_count = 4;
+
+  // The name of the topic that was backed up.
+  string topic_name = 5;
+
+  // The partition count of the topic that was backed up.
+  uint32 partition_count = 6;
+
+  ChecksumAlgorithm checksum_algorithm = 7;
+}
+
+// Each record is wrapped in an envelope in the backup file that contains the
+// checksum of the serialized record. This is used during backup restoration to
+// verify the integrity of each record as it is read.
+message Envelope {
+  Record record = 1;
+
+  // Checksum of the Protobuf serialized record in field 1. The algorithm is
+  // specified in the backup metadata, and a checksum mismatch is going to fail
+  // a backup restoration attempt. Note that this field uses `bytes`, this means
+  // that the algorithm that is used **MUST** produce values with a fixed
+  // endianness across all systems.
+  //
+  // The standard algorithm used by Karapace is xxHash3 64 which always produces
+  // its digest in big endian.
+  bytes checksum = 2;
+}
+
+// The record is a minimal representation of a Kafka record, just enough to
+// restore it. There is no partition index included here, because schema topics
+// always only have a single partition. This is required to ensure perfect
+// ordering of records. Other metadata, like the timestamp, are not used by the
+// schema registry, and can thus be left to their defaults.
+message Record {
+  // The key that uniquely identifies this record. This is JSON for all schema
+  // records and always present. However, keys are optional in Apache Kafka
+  // records.
+  bytes key = 1;
+
+  // The value of this record. This is JSON for all schema records and always
+  // present except for tombstones, hence, this is optional.
+  bytes value = 2;
+
+  // Optional headers for this record. This contains additional metadata for all
+  // schema records.
+  repeated Header headers = 3;
+
+  // The partition the record originated from.
+  uint32 partition = 4;
+
+  // The offset the record had in the backed up topic. This is purely
+  // informational, the record is going to have a new offset after restoration.
+  uint64 offset = 5;
+
+  // The UNIX time the record had in the backed up topic.
+  int64 timestamp = 6;
+
+  // The type of timestamp the record has. At the time of writing valid values
+  // are `-1` for none, `0` for creation, and `1` for log append time.
+  int32 timestamp_type = 7;
```

Comment:
Timestamp type could be an enumeration?

Reply:
I initially used an enum, but an enum requires us to map things (or do some math), because it is not possible to have `-1`. But a simple `timestamp_type.value - 1` would solve that… let‘s do it. 😊